src/parser/mod.rs

```diff
@@ -2266,6 +2312,109 @@ impl<'a> Parser<'a> {
         })
     }
 
+    /// Parse an internal data type of clickhouse nullable [1]
+    /// Internal data type of nullable can’t be any of the composite data types Array, Map and Tuple
+    /// Syntax
+    /// ```sql
+    /// Nullable(data_type)
+    /// ```
+    ///
+    /// [1]: https://clickhouse.com/docs/en/sql-reference/data-types/nullable
+    fn parse_click_house_nullable(&mut self) -> Result<DataType, ParserError> {
+        self.expect_keyword(Keyword::NULLABLE)?;
+
+        self.expect_token(&Token::LParen)?;
+        let (inside_type, _) = self.parse_data_type_helper()?;
+        if matches!(
+            inside_type,
+            DataType::Array(_) | DataType::Map(_, _) | DataType::Tuple(_)
+        ) {
+            return parser_err!("Unexpected internal data type of nullable:", inside_type);
+        }
+        self.expect_token(&Token::RParen)?;
+
+        Ok(inside_type)
+    }
+
+    /// Parse clickhouse low cardinality [1]
+    /// Internal data type of low cardinality can be String, FixedString, Date, DateTime and numbers excepting Decimal
+    /// Syntax
+    /// ```sql
+    /// LowCardinality(data_type)
+    /// ```
+    ///
+    /// [1]: https://clickhouse.com/docs/en/sql-reference/data-types/lowcardinality
+    fn parse_click_house_low_cardinality(&mut self) -> Result<DataType, ParserError> {
+        self.expect_keyword(Keyword::LOWCARDINALITY)?;
+
+        self.expect_token(&Token::LParen)?;
+        let (inside_type, _) = self.parse_data_type_helper()?;
+        if !matches!(
+            inside_type,
+            DataType::String(_) | DataType::Date | DataType::Datetime(_)
+        ) {
+            return parser_err!(
+                "Unexpected internal data type of low cardinality:",
+                inside_type
+            );
+        }
+        self.expect_token(&Token::RParen)?;
+
+        Ok(inside_type)
```

Comment:
I think we can reuse the same functionality used for nullable. If we instead have the type as
```rust
fn parse_sub_type(&mut self, parent_type: F) where F: FnOnce(Box<DataType>) -> DataType {
    self.expect_token(&Token::Paren)?;
    let inside_type = ...;
    self.expect_token(&Token::RParen)?;
    Ok(parent_type(inside_type.into()))
}
```
Then at the call sites, no need to call `self.prev_token()` anymore instead we do e.g.
```rust
                Keyword::LOWCARDINALITY if dialect_of!(self is ClickHouseDialect) => {
                       self.parse_sub_type(DataType::LowCardinality)
                }
```

Reply:
I added a function `parse_sub_type` and updated all related cases.